lib/ex_unit/lib/ex_unit/server.ex

```diff
@@ -91,63 +102,120 @@ defmodule ExUnit.Server do
+
+        updated_state
+        |> update_in([:async_partitions], fn async_partitions ->
+          {_, async_partitions} =
+            Map.get_and_update(async_partitions, async_partition_key, fn
+              nil ->
+                {nil, [name]}
+
+              modules ->
+                {modules, [name | modules]}
+            end)
+
+          async_partitions
+        end)
+        |> update_in([:async_partition_keys], fn q ->
+          if partition_key_exists? do
+            q
+          else
+            :queue.in(async_partition_key, q)
+          end
+        end)
+      end)
+
+    {:reply, :ok, take_module_partitions(state)}
+  end
+
+  def handle_call({:add, {false, _async_partition_key}, names}, _from, %{loaded: loaded} = state)
       when is_integer(loaded) do
     state =
       update_in(state.sync_modules, &Enum.reduce(names, &1, fn name, q -> :queue.in(name, q) end))
 
     {:reply, :ok, state}
   end
 
-  def handle_call({:add, _async?, _names}, _from, state),
+  def handle_call({:add, {_async?, _async_partition_key}, _names}, _from, state),
     do: {:reply, :already_running, state}
 
-  defp take_modules(%{waiting: nil} = state) do
+  defp default_async_partition_key({module, params}) do
+    if params == %{} do
+      module
+    else
+      # if no async partition is specified, parameters run concurrently
+      "#{module}_#{:erlang.phash2(params)}"
+    end
+  end
+
+  defp take_module_partitions(%{waiting: nil} = state) do
     state
   end
 
-  defp take_modules(%{waiting: {from, count}} = state) do
-    has_async_modules? = not :queue.is_empty(state.async_modules)
+  defp take_module_partitions(%{waiting: {from, count}} = state) do
+    has_async_partitions? = not :queue.is_empty(state.async_partition_keys)
 
     cond do
-      not has_async_modules? and state.loaded == :done ->
+      not has_async_partitions? and state.loaded == :done ->
         GenServer.reply(from, nil)
         %{state | waiting: nil}
 
-      not has_async_modules? ->
+      not has_async_partitions? ->
         state
 
       true ->
-        {modules, async_modules} = take_until(count, state.async_modules)
-        GenServer.reply(from, modules)
-        %{state | async_modules: async_modules, waiting: nil}
+        {partition_keys, remaining_partition_keys} =
+          take_until(count, state.async_partition_keys)
+
+        {async_partitions, remaining_partitions} =
+          Enum.map_reduce(partition_keys, state.async_partitions, fn partition_key,
+                                                                     remaining_partitions ->
+            {partition_modules, remaining_partitions} =
+              Map.pop!(remaining_partitions, partition_key)
+
+            {
+              {partition_key, Enum.reverse(partition_modules)},
+              remaining_partitions
+            }
+          end)
+
+        GenServer.reply(from, async_partitions)
```

Comment:
Apologies but I may be missing something here: how do you avoid overlapping tests in the same group if the server does not know which partitions are currently running? Wouldn't you need the client/runner to tell you which partitions are currently available and which ones are not?

Reply:
With the proposed changes, the runner take all the modules in a group and executes them in order:

```elixir
{pid, ref} =
  spawn_monitor(fn ->
    Enum.each(partition_modules, fn {module, params} ->
      run_module(config, module, async?, params)
    end)
  end)
```

Do you see an issue with that approach?

Reply:
This is fine after all: https://github.com/elixir-lang/elixir/pull/13897#discussion_r1798214835